Easier selection of variables in Radviz, Linear projection and similar

janezd opened the pull request:
##### Issue

We had two listviews which took too much space. So we switched to +/- buttons, which are better require more clicks and are somewhat inconsistent (to remove, you select and click -, while to add you click + and select) although similar to other dialogs (at least on macOS).

##### Description of changes

I propose a single list view in which the user can add and remove. I recorded three alternative visual indications; other suggestions welcome. The user can also drag a variable from shown to hidden or the opposite, which I forgot to record.

I don't particularly like the second one. I lean towards the third.

(Yes, I know, the indicator does not disappear when mouse leaves the view. It will. Settings and tests are also broken.)

Below the list view, we can add buttons Clear and Reset.

![radviz1](https://user-images.githubusercontent.com/2387315/50558968-913fb080-0cf2-11e9-9d76-ea500cf89c52.gif)

![radviz2](https://user-images.githubusercontent.com/2387315/50558905-0c549700-0cf2-11e9-87db-4454fe18fa00.gif)

![radviz3](https://user-images.githubusercontent.com/2387315/50558909-14143b80-0cf2-11e9-8e11-be92388660a9.gif)

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

markotoplak commented:
The second one is my favorite. It is the most explicit, which is probably important for such non-standard UI. It is also non-obvious that white are selected and grey are not (here the text helped a lot).

I would prefer using standard system selected/unselected colors. Selected is usually not white.

I also like that the 2nd does not introduce any new colors into the UI. The colored variants look strange and might look even stranger on systems with other color schemas.

ajdapretnar commented:
Sorry for adding the fourth option, but you should probably check the Timeseries add on, specifically the Line Chart widget. However, the implementation there is not really done well. Specifically on OSX, selection does not finish upon clicking on an attribute, instead it gets dragged around selecting attributes on its way. You have to double click the attribute for the selection to stay put... 

<img width="907" alt="screen shot 2019-01-11 at 10 32 30" src="https://user-images.githubusercontent.com/12524972/51025556-ae933b00-158c-11e9-99cf-78bda39cba45.png">

BlazZupan commented:
I agree with Marko, I like the second variant -- which is most explicit -- best.

janezd commented:
Make the line stronger.

Add line below or above the group if none or all shown variables are selected (e.g. when filtering).

Change the number of selected variables when dragging variables into or out of selection.

markotoplak commented:
I checked this, works well except for the dragging. I like how it looks, but yes, the line should be shown also if one of the groups is empty.

VesnaT commented on Orange/widgets/utils/plot/owplotgui.py:
```diff
@@ -51,165 +53,175 @@
+        proxy.invalidate()
+        view.selectionModel().clearSelection()
+
+    filter, view = variables_filter(
```
`filter` is not my favourite choice of a variable name since it is a built-in function. How about `filter_edit` as in the `variables_filter`?

janezd replied:
Mine neither. Pylint's neither. Changed to `filter_edit`.

janezd commented:
> I would prefer more space between the top and the Filer box

After realizing what you meant by "Filer", I added some space. I however think the space was the same as before.

> once a variable is selected and not placed at the end of the selected variables, it is impossible to place it at the end. 

I fixed this, too. Also, if you drag an unselected variable just to between the last selected and the first unselected, the dragged variable is added to selection. Or if the selection is empty, dragging to the top adds the variable (not the most useful functionality, though - clicking is simpler).

janezd commented on Orange/widgets/utils/plot/owplotgui.py:
```diff
@@ -51,165 +53,175 @@
+
+    def __init__(self, selected_vars):
+        super().__init__(enable_dnd=True)
+        self.selected_vars = selected_vars
```
It's a list of `Variable`. That's a part of the whole point. :)

It was thought that settings would keep being compatible, but I later implemented settings differently, so you're right. Fixed.

janezd commented on Orange/widgets/visualize/owlinearprojection.py:
```diff
@@ -322,19 +311,18 @@ def continuous_variables(self):
+        # the view will have to connect
+        self.model_selected.selection_changed.emit()
         self.__model_selected_changed()
 
```
Indeed. Good catch.

janezd commented on Orange/widgets/visualize/owradviz.py:
```diff
@@ -47,7 +46,7 @@ def __init__(self, master):
         self.master = master
         self.n_neighbors = 10
-        max_n_attrs = len(master.model_selected) + len(master.model_other) - 1
+        max_n_attrs = 0
```
It may had been meant as just a temporary fix because the code crashed after eliminating `model_selected` and `model_other`. This however never worked, I suppose: Vizrank's `__init__` is called from widget's `__init__` not from `set_data`, so the number of attributes is unknown, and also not changed when the data changes.

I changed the max to 99. If the number of attributes is smaller, it's OK. If it's larger, this limit is sensible, too.
